### ISSUE TITLE
Opcode selection

### ISSUE THREAD

**Issue by snoopy-coder** [x64dbg/x64dbg#3395]
### Feature type

Other

### Describe the feature

Some x86 instructions have two opcodes (x64dbg seems to choose the first one in order by default), Is it difficult to add the option to select the second opcode manually (switching between r/rm and rm/r)? Thanks a lot. 

![x64dbg-asm](https://github.com/user-attachments/assets/a0248400-d292-4f82-bb6d-1d33fc6716ff)

**Comment by mrexodia** [comment]
This functionality is not exposed by the https://github.com/asmjit/asmtk/ project used by this dialog. I cannot find the old discussion, but basically asmjit will always choose the shortest opcode. It might also emit `xor eax, eax` instead of `xor rax, rax` for example.

**Comment by snoopy-coder** [comment]
I am not an expert in this matter and found only a couple of closed issues mentioning that options kX86_ModMR and  kX86_ModRM have been added some time ago:

[https://github.com/asmjit/asmjit/issues/64](https://github.com/asmjit/asmjit/issues/64)
[https://github.com/asmjit/asmjit/issues/392](https://github.com/asmjit/asmjit/issues/392)

They are also listed in the [API Reference](https://asmjit.com/doc/group__asmjit__instruction__db.html#ga3cb186f79692703d2073e0dc8dfdb473) on asmjit.com

**Comment by mrexodia** [comment]
Yes, but x64dbg does not use asmjit directly. It uses the asmtk project and I don't think it's exposed there. If you're interested to work on this you would have to modify https://github.com/x64dbg/asmjit_xedparse/blob/master/asmjit/XEDParseEmulator.cpp#L41 and it would also need special GUI support. For me personally this is not worth investigating, but I will reopen the issue in case someone else wants to pick it up.

**Comment by kobalicek** [comment]
AsmTK supports the selection, the interactive parser (https://asmjit.com/parser.html) can be used to verify:

For example you can try this:

```
mov eax, ebx
modmr mov eax, ebx
modrm mov eax, ebx
```

The reason both modmr and modrm are exposed is because AsmJit tries to be binary compatible with LLVM, which isn't consistent in some cases (sometimes preferring MODRM and sometimes MODMR).

The case with xor - `xor eax, eax` and `xor rax, rax` should always be encoded differently (xor rax, rax would use REX prefix).

**Comment by mrexodia** [comment]
Oh awesome, thanks for the information!

**Comment by snoopy-coder** [comment]
> AsmTK supports the selection
Thanks a lot, Petr!
Indeed, the command `modrm xor eax, eax` works as expected and outputs `33c0`
By default, or `modmr xor eax, eax` outputs `31c0`
![asm_modrm](https://github.com/user-attachments/assets/235a6aee-7c80-445e-a052-0db26fd4263e)
